integration/test/Testlib/App.hs

```diff
@@ -0,0 +1,935 @@
+    "pQABARn//wKhAFgg5ER8h0/bIADXjBXe/XPKdzekgv6nhJ4hp3vJ3jtTSbUDoQChAFgglacihnqg/YQJHkuHNFU7QD6Pb3KN4FnubaCF2EVOgRkE9g==",
+    "pQABARn//wKhAFggsgV6jq+GuNuvXk+ctHh570cNqEmfPhz34wcYCMCf9xIDoQChAFgglacihnqg/YQJHkuHNFU7QD6Pb3KN4FnubaCF2EVOgRkE9g==",
+    "pQABARn//wKhAFggdQdlPqkBw6+phKhohp3YaWQL710euZDnyMLFwf2cS0oDoQChAFgglacihnqg/YQJHkuHNFU7QD6Pb3KN4FnubaCF2EVOgRkE9g==",
+    "pQABARn//wKhAFggKlsI/snuQMoYcZRw/kN+BobPV5gwYeBClp0Wx9btTGUDoQChAFgglacihnqg/YQJHkuHNFU7QD6Pb3KN4FnubaCF2EVOgRkE9g==",
+    "pQABARn//wKhAFggtruFBClEgdPKvjpHsYLlWMev9L4OmYZwlxbY0NwvzOwDoQChAFgglacihnqg/YQJHkuHNFU7QD6Pb3KN4FnubaCF2EVOgRkE9g==",
+    "pQABARn//wKhAFggRUdh4cuYtFNL46RLnPy65goYInyreStKwsEcY3pPlLkDoQChAFgglacihnqg/YQJHkuHNFU7QD6Pb3KN4FnubaCF2EVOgRkE9g==",
+    "pQABARn//wKhAFggQtT7lLZzH171F4jCbHNwxEAt28FwdQ8Kt2tbxFzPgC0DoQChAFgglacihnqg/YQJHkuHNFU7QD6Pb3KN4FnubaCF2EVOgRkE9g==",
+    "pQABARn//wKhAFggQeUPM119c+6zRsEupA8zshTfrZiLpXx1Ji0UMMumq9IDoQChAFgglacihnqg/YQJHkuHNFU7QD6Pb3KN4FnubaCF2EVOgRkE9g=="
+  ]
+
+-------------------------------------------------------------------------------
+-- - SECTION_ASSERTIONS
+-------------------------------------------------------------------------------
+
+data AssertionFailure = AssertionFailure
+  { callstack :: CallStack,
+    response :: Maybe Response,
+    msg :: String
+  }
+
+instance Show AssertionFailure where
+  show (AssertionFailure _ _ msg) = "AssertionFailure _ _ " <> show msg
+
+instance Exception AssertionFailure where
+  displayException (AssertionFailure _ _ msg) = msg
+
+assertFailure :: HasCallStack => String -> App a
+assertFailure msg =
+  deepseq msg $
+    liftIO $
+      E.throw (AssertionFailure callStack Nothing msg)
+
+assertBool :: HasCallStack => String -> Bool -> App ()
+assertBool _ True = pure ()
+assertBool msg False = assertFailure msg
+
+assertOne :: HasCallStack => [a] -> App a
+assertOne [x] = pure x
+assertOne xs = assertFailure ("Expected one, but got " <> show (length xs))
+
+expectFailure :: HasCallStack => (AssertionFailure -> App ()) -> App a -> App ()
+expectFailure checkFailure action = do
+  env <- ask
+  res :: Either AssertionFailure x <-
+    liftIO
+      (E.try (runAppWithEnv env action))
+  case res of
+    Left e@(AssertionFailure callstack mr msg) ->
+      checkFailure e
+    Right x -> assertFailure "Expected AssertionFailure, but none occured"
+
+shouldMatch ::
+  (ProducesJSON a, ProducesJSON b, HasCallStack) =>
+  -- | The actual value
+  a ->
+  -- | The expected value
+  b ->
+  App ()
+a `shouldMatch` b = do
+  xa <- prodJSON a
+  xb <- prodJSON b
+  unless (xa == xb) $ do
+    pa <- prettyJSON xa
+    pb <- prettyJSON xb
+    assertFailure $ "Expected:\n" <> pb <> "\n" <> "Actual:\n" <> pa
+
+shouldMatchInt ::
+  (ProducesJSON a, HasCallStack) =>
+  -- | The actual value
+  a ->
+  -- | The expected value
+  Int ->
+  App ()
+a `shouldMatchInt` b = do
+  xa <- prodJSON a
+  shouldMatch xa b
```

Comment:
```suggestion
-- | Specialized variant of `shouldMatch` to avoid the need for type annotations.
shouldMatchInt ::
  (ProducesJSON a, HasCallStack) =>
  -- | The actual value
  a ->
  -- | The expected value
  Int ->
  App ()
shouldMatchInt = shouldMatch

```

seems to compile, and it makes sense to me...

Reply:
Nice. Applied